File: general-solving/leetcode/102_binary-tree-level-order-traversal.cs
/***************************************************************************
* Problem Name: Binary Tree Level Order Traversal
* Problem URL : https://leetcode.com/problems/binary-tree-level-order-traversal/
* Date        : Jan 5 2015
* Complexity  : O(n) Time
* Author      : Atiq Rahman
* Status      : Accepted (405ms)
* Notes       : consider that nodes of each level stays on specific index on
*               the vector of vectors
*               this is the C# version from the cpp solution accepted first
* meta        : tag-binary-tree
***************************************************************************/

public class Solution {
    IList<IList<int>> Node_list_list;
    public IList<IList<int>> LevelOrder(TreeNode root) {
        Node_list_list = new List<IList<int>>();
        LevelOrderRec(root, 0);
        return Node_list_list;
    }
    
    void LevelOrderRec(TreeNode root, int index) {
        if (root == null)
            return ;
        if (Node_list_list.Count <= index) {
            IList<int> temp = new List<int>();
            Node_list_list.Add(temp);
        }
        Node_list_list[index].Add(root.val);
        LevelOrderRec(root.left, index + 1);
        LevelOrderRec(root.right, index + 1);
    }
}

File: general-solving/leetcode/102_binary-tree-level-order-traversal.cpp
/***************************************************************************
* Problem Name: Binary Tree Level Order Traversal
* Problem URL : https://leetcode.com/problems/binary-tree-level-order-traversal/
* Date        : Jan 5 2015
* Complexity  : O(n) Time
* Author      : Atiq Rahman
* Status      : Accepted
* Notes       : consider that nodes of each level stays on specific index on the vector of vectors
* meta        : tag-binary-tree
***************************************************************************/

class Solution {
private:
    vector<vector<int>> tr_node_list;
    
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        levelOrderRec(root, 0);
        return tr_node_list;
    }
    void levelOrderRec(TreeNode* root, int index) {
        if (root == NULL)
            return;
        if (tr_node_list.size() <= index) {
            vector<int> temp;
            tr_node_list.push_back(temp);
        }
        tr_node_list[index].push_back(root->val);
        levelOrderRec(root->left, index+1);
        levelOrderRec(root->right, index+1);
    }
};
